man/dpkg-maintscript-helper.1
.\" dpkg manual page - dpkg-maintscript-helper(1)
.\"
.\" Copyright © 2010-2012 Raphaël Hertzog <hertzog@debian.org>
.\" Copyright © 2011-2015 Guillem Jover <guillem@debian.org>
.\"
.\" This is free software; you can redistribute it and/or modify
.\" it under the terms of the GNU General Public License as published by
.\" the Free Software Foundation; either version 2 of the License, or
.\" (at your option) any later version.
.\"
.\" This is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public License
.\" along with this program.  If not, see <https://www.gnu.org/licenses/>.
.
.TH dpkg\-maintscript\-helper 1 "2014-09-01" "Debian Project" "dpkg suite"
.SH NAME
dpkg\-maintscript\-helper \- works around known dpkg limitations in maintainer scripts
.
.SH SYNOPSIS
.B dpkg\-maintscript\-helper
.IR command " [" parameter "...] \fB\-\-\fP " maint-script-parameter ...
.
.SH COMMANDS AND PARAMETERS
.P
\fBsupports\fP \fIcommand\fP
.P
\fBrm_conffile\fP \fIconffile\fP [\fIprior-version\fP [\fIpackage\fP]]
.P
\fBmv_conffile\fP \fIold-conffile\fP \fInew-conffile\fP [\fIprior-version\fP [\fIpackage\fP]]
.P
\fBsymlink_to_dir\fP \fIpathname\fP \fIold-target\fP [\fIprior-version\fP [\fIpackage\fP]]
.P
\fBdir_to_symlink\fP \fIpathname\fP \fInew-target\fP [\fIprior-version\fP [\fIpackage\fP]]
.
.SH DESCRIPTION
.P
This program is designed to be run within maintainer scripts to achieve
some tasks that \fBdpkg\fP can't (yet) handle natively either because of
design decisions or due to current limitations.
.P
Many of those tasks require coordinated actions from several maintainer
scripts (\fBpreinst\fP, \fBpostinst\fP, \fBprerm\fP, \fBpostrm\fP). To
avoid mistakes the same call simply needs to be put in all scripts and the
program will automatically adapt its behaviour based on the environment
variable \fBDPKG_MAINTSCRIPT_NAME\fP and on the maintainer scripts arguments
that you have to forward after a double hyphen.
.
.SH COMMON PARAMETERS
.TP
.I prior-version
Defines the latest version of the package whose upgrade should trigger the
operation. It is important to calculate \fIprior-version\fP correctly so
that the operations are correctly performed even if the user rebuilt the
package with a local version. If \fIprior-version\fP is empty or omitted,
then the operation is tried on every upgrade (note: it's safer to give
the version and have the operation tried only once).

If the conffile has not been shipped for several versions, and you are
now modifying the maintainer scripts to clean up the obsolete file,
\fIprior-version\fP should be based on the version of the package that
you are now preparing, not the first version of the package that lacked
the conffile. This applies to all other actions in the same way.

For example, for a conffile removed in version \fB2.0\-1\fP of a package,
\fIprior-version\fP should be set to \fB2.0\-1~\fP. This will cause the
conffile to be removed even if the user rebuilt the previous version
\fB1.0\-1\fP as \fB1.0\-1local1\fP. Or a package switching a path from
a symlink (shipped in version \fB1.0\-1\fP) to a directory (shipped in
version \fB2.0\-1\fP), but only performing the actual switch in the
maintainer scripts in version \fB3.0\-1\fP, should set \fIprior-version\fP
to \fB3.0\-1~\fP.
.TP
.I package
The package name. When the package is \(lqMulti\-Arch: same\(rq this parameter
must include the architecture qualifier, otherwise it should \fBnot\fP
usually include the architecture qualifier (as it would disallow
cross-grades, or switching from being architecture specific to
architecture \fBall\fP or vice versa).
If the parameter is empty or omitted, the \fBDPKG_MAINTSCRIPT_PACKAGE\fP
and \fBDPKG_MAINTSCRIPT_ARCH\fP environment variables (as set by \fBdpkg\fP)
will be used to generate an arch-qualified package name.
.TP
.B \-\-
All the parameters of the maintainer scripts have to be forwarded to the
program after \fB\-\-\fP.
.SH CONFFILE RELATED TASKS
.P
When upgrading a package, \fBdpkg\fP will not automatically remove a conffile
(a configuration file for which \fBdpkg\fP should preserve user changes) if
it is not present in the newer version. There are two principal reasons for
this; the first is that the conffile could've been dropped by accident and
the next version could restore it, users wouldn't want their changes
thrown away. The second is to allow packages to transition files from a
dpkg\-maintained conffile to a file maintained by the package's maintainer
scripts, usually with a tool like debconf or ucf.
.P
This means that if a package is intended to rename or remove a conffile,
it must explicitly do so and \fBdpkg\-maintscript\-helper\fP can be used
to implement graceful deletion and moving of conffiles within maintainer
scripts.
.
.SS Removing a conffile
.P
If a conffile is completely removed, it should be removed from disk,
unless the user has modified it. If there are local modifications, they
should be preserved. If the package upgrades aborts, the newly obsolete
conffile should not disappear.
.P
All of this is implemented by putting the following shell snippet in the
\fBpreinst\fP, \fBpostinst\fP and \fBpostrm\fP maintainer scripts:
.P
    dpkg\-maintscript\-helper rm_conffile \\
        \fIconffile\fP \fIprior-version\fP \fIpackage\fP \-\- "$@"
.P
\fIconffile\fP is the filename of the conffile to remove.
.P
Current implementation: in the \fBpreinst\fP, it checks if the conffile
was modified and renames it either to \fIconffile\fP\fB.dpkg\-remove\fP (if not
modified) or to \fIconffile\fP\fB.dpkg\-backup\fP (if modified). In the
\fBpostinst\fP, the latter file is renamed to \fIconffile\fP\fB.dpkg\-bak\fP
and kept for reference as it contains user modifications but the former will
be removed. If the package upgrade aborts, the \fBpostrm\fP reinstalls the
original conffile. During purge, the \fBpostrm\fP will also delete the
\fB.dpkg\-bak\fP file kept up to now.
.
.SS Renaming a conffile
.P
If a conffile is moved from one location to another, you need to make sure
you move across any changes the user has made. This may seem a simple
change to the \fBpreinst\fP script at first, however that will result in
the user being prompted by \fBdpkg\fP to approve the conffile edits even
though they are not responsible of them.
.P
Graceful renaming can be implemented by putting the following shell
snippet in the \fBpreinst\fP, \fBpostinst\fP and \fBpostrm\fP maintainer
scripts:
.P
    dpkg\-maintscript\-helper mv_conffile \\
        \fIold-conffile\fP \fInew-conffile\fP \fIprior-version\fP \fIpackage\fP \-\- "$@"
.P
\fIold-conffile\fP and \fInew-conffile\fP are the old and new name of the
conffile to rename.
.P
Current implementation: the \fBpreinst\fP checks if the conffile has been
modified, if yes it's left on place otherwise it's renamed to
\fIold-conffile\fP\fB.dpkg\-remove\fP. On configuration, the \fBpostinst\fP
removes \fIold-conffile\fP\fB.dpkg\-remove\fP and renames \fIold-conffile\fP
to \fInew-conffile\fP if \fIold-conffile\fP is still available. On
abort\-upgrade/abort\-install, the \fBpostrm\fP renames
\fIold-conffile\fP\fB.dpkg\-remove\fP back to \fIold-conffile\fP if required.
.
.SH SYMLINK AND DIRECTORY SWITCHES
.
When upgrading a package, \fBdpkg\fP will not automatically switch a symlink
to a directory or vice-versa. Downgrades are not supported and the path
will be left as is.
.
.SS Switching a symlink to directory
.
If a symlink is switched to a real directory, you need to make sure
before unpacking that the symlink is removed. This may seem a simple
change to the \fBpreinst\fP script at first, however that will result
in some problems in case of admin local customization of the symlink
or when downgrading the package.
.P
Graceful renaming can be implemented by putting the following shell
snippet in the \fBpreinst\fP, \fBpostinst\fP and \fBpostrm\fP maintainer
scripts:
.P
    dpkg\-maintscript\-helper symlink_to_dir \\
        \fIpathname\fP \fIold-target\fP \fIprior-version\fP \fIpackage\fP \-\- "$@"
.P
\fIpathname\fP is the absolute name of the old symlink (the path will be a
directory at the end of the installation) and \fIold-target\fP is
the target name of the former symlink at \fIpathname\fP. It can either be
absolute or relative to the directory containing \fIpathname\fP.
.P
Current implementation: the \fBpreinst\fP checks if the symlink exists
and points to \fIold-target\fP, if not then it's left in place, otherwise
it's renamed to \fIpathname\fP\fB.dpkg\-backup\fP. On configuration,
the \fBpostinst\fP removes \fIpathname\fP\fB.dpkg\-backup\fP if
\fIpathname\fP\fB.dpkg\-backup\fP is still a symlink. On
abort\-upgrade/abort\-install, the \fBpostrm\fP renames
\fIpathname\fP\fB.dpkg\-backup\fP back to \fIpathname\fP if required.
.
.SS Switching a directory to symlink
.
If a real directory is switched to a symlink, you need to make sure
before unpacking that the directory is removed. This may seem a simple
change to the \fBpreinst\fP script at first, however that will result
in some problems in case the directory contains conffiles, pathnames
owned by other packages, locally created pathnames, or when downgrading
the package.
.P
Graceful switching can be implemented by putting the following shell
snippet in the \fBpreinst\fP, \fBpostinst\fP and \fBpostrm\fP maintainer
scripts:
.P
    dpkg\-maintscript\-helper dir_to_symlink \\
        \fIpathname\fP \fInew-target\fP \fIprior-version\fP \fIpackage\fP \-\- "$@"
.P
\fIpathname\fP is the absolute name of the old directory (the path
will be a symlink at the end of the installation) and \fInew-target\fP is
the target of the new symlink at \fIpathname\fP. It can either be absolute
or relative to the directory containing \fIpathname\fP.
.P
Current implementation: the \fBpreinst\fP checks if the directory
exists, does not contain conffiles, pathnames owned by other packages,
or locally created pathnames, if not then it's left in place, otherwise
it's renamed to \fIpathname\fP\fB.dpkg\-backup\fP, and an empty staging
directory named \fIpathname\fP is created, marked with a file so that
dpkg can track it. On configuration, the \fBpostinst\fP finishes the
switch if \fIpathname\fP\fB.dpkg\-backup\fP is still a directory and
\fIpathname\fP is the staging directory; it removes the staging directory
mark file, moves the newly created files inside the staging directory
to the symlink target \fInew-target\fP/, replaces the now empty staging
directory \fIpathname\fP with a symlink to \fInew-target\fP, and
removes \fIpathname\fP\fB.dpkg\-backup\fP. On
abort\-upgrade/abort\-install, the \fBpostrm\fP renames

scripts/dpkg-maintscript-helper.sh
#!/bin/sh
#
# Copyright © 2007, 2011-2015 Guillem Jover <guillem@debian.org>
# Copyright © 2010 Raphaël Hertzog <hertzog@debian.org>
# Copyright © 2008 Joey Hess <joeyh@debian.org>
# Copyright © 2005 Scott James Remnant (original implementation on www.dpkg.org)
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <https://www.gnu.org/licenses/>.

# The conffile related functions are inspired by
# https://wiki.debian.org/DpkgConffileHandling

# This script is documented in dpkg-maintscript-helper(1)

##
## Functions to remove an obsolete conffile during upgrade
##
rm_conffile() {
	local CONFFILE="$1"
	local LASTVERSION="$2"
	local PACKAGE="$3"
	if [ "$LASTVERSION" = "--" ]; then
		LASTVERSION=""
		PACKAGE="$DPKG_MAINTSCRIPT_PACKAGE${DPKG_MAINTSCRIPT_ARCH:+:$DPKG_MAINTSCRIPT_ARCH}"
	fi
	if [ "$PACKAGE" = "--" -o -z "$PACKAGE" ]; then
		PACKAGE="$DPKG_MAINTSCRIPT_PACKAGE${DPKG_MAINTSCRIPT_ARCH:+:$DPKG_MAINTSCRIPT_ARCH}"
	fi
	# Skip remaining parameters up to --
	while [ "$1" != "--" -a $# -gt 0 ]; do shift; done
	[ $# -gt 0 ] || badusage "missing arguments after --"
	shift

	[ -n "$PACKAGE" ] || error "couldn't identify the package"
	[ -n "$1" ] || error "maintainer script parameters are missing"
	[ -n "$DPKG_MAINTSCRIPT_NAME" ] || \
		error "environment variable DPKG_MAINTSCRIPT_NAME is required"

	debug "Executing $0 rm_conffile in $DPKG_MAINTSCRIPT_NAME" \
	      "of $DPKG_MAINTSCRIPT_PACKAGE"
	debug "CONFFILE=$CONFFILE PACKAGE=$PACKAGE" \
	      "LASTVERSION=$LASTVERSION ACTION=$1 PARAM=$2"
	case "$DPKG_MAINTSCRIPT_NAME" in
	preinst)
		if [ "$1" = "install" -o "$1" = "upgrade" ] && [ -n "$2" ] &&
		   dpkg --compare-versions "$2" le-nl "$LASTVERSION"; then
			prepare_rm_conffile "$CONFFILE" "$PACKAGE"
		fi
		;;
	postinst)
		if [ "$1" = "configure" ] && [ -n "$2" ] &&
		   dpkg --compare-versions "$2" le-nl "$LASTVERSION"; then
			finish_rm_conffile "$CONFFILE"
		fi
		;;
	postrm)
		if [ "$1" = "purge" ]; then
			rm -f "$CONFFILE.dpkg-bak" "$CONFFILE.dpkg-remove" \
			      "$CONFFILE.dpkg-backup"
		fi
		if [ "$1" = "abort-install" -o "$1" = "abort-upgrade" ] &&
		   [ -n "$2" ] &&
		   dpkg --compare-versions "$2" le-nl "$LASTVERSION"; then
			abort_rm_conffile "$CONFFILE" "$PACKAGE"
		fi
		;;
	*)
		debug "$0 rm_conffile not required in $DPKG_MAINTSCRIPT_NAME"
		;;
	esac
}

prepare_rm_conffile() {
	local CONFFILE="$1"
	local PACKAGE="$2"

	[ -e "$CONFFILE" ] || return 0
	ensure_package_owns_file "$PACKAGE" "$CONFFILE" || return 0

	local md5sum="$(md5sum $CONFFILE | sed -e 's/ .*//')"
	local old_md5sum="$(dpkg-query -W -f='${Conffiles}' $PACKAGE | \
		sed -n -e "\' $CONFFILE ' { s/ obsolete$//; s/.* //; p }")"
	if [ "$md5sum" != "$old_md5sum" ]; then
		mv -f "$CONFFILE" "$CONFFILE.dpkg-backup"
	else
		mv -f "$CONFFILE" "$CONFFILE.dpkg-remove"
	fi
}

finish_rm_conffile() {
	local CONFFILE="$1"

	if [ -e "$CONFFILE.dpkg-backup" ]; then
		echo "Obsolete conffile $CONFFILE has been modified by you."
		echo "Saving as $CONFFILE.dpkg-bak ..."
		mv -f "$CONFFILE.dpkg-backup" "$CONFFILE.dpkg-bak"
	fi
	if [ -e "$CONFFILE.dpkg-remove" ]; then
		echo "Removing obsolete conffile $CONFFILE ..."
		rm -f "$CONFFILE.dpkg-remove"
	fi
}

abort_rm_conffile() {
	local CONFFILE="$1"
	local PACKAGE="$2"

	ensure_package_owns_file "$PACKAGE" "$CONFFILE" || return 0

	if [ -e "$CONFFILE.dpkg-remove" ]; then
		echo "Reinstalling $CONFFILE that was moved away"
		mv "$CONFFILE.dpkg-remove" "$CONFFILE"
	fi
	if [ -e "$CONFFILE.dpkg-backup" ]; then
		echo "Reinstalling $CONFFILE that was backupped"
		mv "$CONFFILE.dpkg-backup" "$CONFFILE"
	fi
}

##
## Functions to rename a conffile during upgrade
##
mv_conffile() {
	local OLDCONFFILE="$1"
	local NEWCONFFILE="$2"
	local LASTVERSION="$3"
	local PACKAGE="$4"
	if [ "$LASTVERSION" = "--" ]; then
		LASTVERSION=""
		PACKAGE="$DPKG_MAINTSCRIPT_PACKAGE${DPKG_MAINTSCRIPT_ARCH:+:$DPKG_MAINTSCRIPT_ARCH}"
	fi
	if [ "$PACKAGE" = "--" -o -z "$PACKAGE" ]; then
		PACKAGE="$DPKG_MAINTSCRIPT_PACKAGE${DPKG_MAINTSCRIPT_ARCH:+:$DPKG_MAINTSCRIPT_ARCH}"
	fi
	# Skip remaining parameters up to --
	while [ "$1" != "--" -a $# -gt 0 ]; do shift; done
	[ $# -gt 0 ] || badusage "missing arguments after --"
	shift

	[ -n "$PACKAGE" ] || error "couldn't identify the package"
	[ -n "$1" ] || error "maintainer script parameters are missing"
	[ -n "$DPKG_MAINTSCRIPT_NAME" ] || \
		error "environment variable DPKG_MAINTSCRIPT_NAME is required"

	debug "Executing $0 mv_conffile in $DPKG_MAINTSCRIPT_NAME" \
	      "of $DPKG_MAINTSCRIPT_PACKAGE"
	debug "CONFFILE=$OLDCONFFILE -> $NEWCONFFILE PACKAGE=$PACKAGE" \
	      "LASTVERSION=$LASTVERSION ACTION=$1 PARAM=$2"
	case "$DPKG_MAINTSCRIPT_NAME" in
	preinst)
		if [ "$1" = "install" -o "$1" = "upgrade" ] && [ -n "$2" ] &&
		   dpkg --compare-versions "$2" le-nl "$LASTVERSION"; then
			prepare_mv_conffile "$OLDCONFFILE" "$PACKAGE"
		fi
		;;
	postinst)
		if [ "$1" = "configure" ] && [ -n "$2" ] &&
		   dpkg --compare-versions "$2" le-nl "$LASTVERSION"; then
			finish_mv_conffile "$OLDCONFFILE" "$NEWCONFFILE" "$PACKAGE"
		fi
		;;
	postrm)
		if [ "$1" = "abort-install" -o "$1" = "abort-upgrade" ] &&
		   [ -n "$2" ] &&
		   dpkg --compare-versions "$2" le-nl "$LASTVERSION"; then
			abort_mv_conffile "$OLDCONFFILE" "$PACKAGE"
		fi
		;;
	*)
		debug "$0 mv_conffile not required in $DPKG_MAINTSCRIPT_NAME"
		;;
	esac
}

prepare_mv_conffile() {
	local CONFFILE="$1"
	local PACKAGE="$2"

	[ -e "$CONFFILE" ] || return 0

	ensure_package_owns_file "$PACKAGE" "$CONFFILE" || return 0

	local md5sum="$(md5sum $CONFFILE | sed -e 's/ .*//')"
	local old_md5sum="$(dpkg-query -W -f='${Conffiles}' $PACKAGE | \
		sed -n -e "\' $CONFFILE ' { s/ obsolete$//; s/.* //; p }")"
	if [ "$md5sum" = "$old_md5sum" ]; then
		mv -f "$CONFFILE" "$CONFFILE.dpkg-remove"
	fi
}

finish_mv_conffile() {
	local OLDCONFFILE="$1"
	local NEWCONFFILE="$2"
	local PACKAGE="$3"

	rm -f $OLDCONFFILE.dpkg-remove

	[ -e "$OLDCONFFILE" ] || return 0
	ensure_package_owns_file "$PACKAGE" "$OLDCONFFILE" || return 0

	echo "Preserving user changes to $NEWCONFFILE (renamed from $OLDCONFFILE)..."
	if [ -e "$NEWCONFFILE" ]; then
		mv -f "$NEWCONFFILE" "$NEWCONFFILE.dpkg-new"
	fi
	mv -f "$OLDCONFFILE" "$NEWCONFFILE"
}

abort_mv_conffile() {
	local CONFFILE="$1"
	local PACKAGE="$2"

	ensure_package_owns_file "$PACKAGE" "$CONFFILE" || return 0

	if [ -e "$CONFFILE.dpkg-remove" ]; then
		echo "Reinstalling $CONFFILE that was moved away"
		mv "$CONFFILE.dpkg-remove" "$CONFFILE"
	fi
}

##
## Functions to replace a symlink with a directory
##
symlink_to_dir() {
	local SYMLINK="$1"
	local SYMLINK_TARGET="$2"
	local LASTVERSION="$3"
	local PACKAGE="$4"

	if [ "$LASTVERSION" = "--" ]; then
		LASTVERSION=""
		PACKAGE="$DPKG_MAINTSCRIPT_PACKAGE${DPKG_MAINTSCRIPT_ARCH:+:$DPKG_MAINTSCRIPT_ARCH}"
	fi
	if [ "$PACKAGE" = "--" -o -z "$PACKAGE" ]; then
		PACKAGE="$DPKG_MAINTSCRIPT_PACKAGE${DPKG_MAINTSCRIPT_ARCH:+:$DPKG_MAINTSCRIPT_ARCH}"
	fi

	# Skip remaining parameters up to --
	while [ "$1" != "--" -a $# -gt 0 ]; do shift; done
	[ $# -gt 0 ] || badusage "missing arguments after --"
	shift

	[ -n "$DPKG_MAINTSCRIPT_NAME" ] || \
		error "environment variable DPKG_MAINTSCRIPT_NAME is required"
	[ -n "$PACKAGE" ] || error "cannot identify the package"
	[ -n "$SYMLINK" ] || error "symlink parameter is missing"
	[ "${SYMLINK#/}" = "$SYMLINK" ] && \
		error "symlink pathname is not an absolute path"
	[ "${SYMLINK%/}" = "$SYMLINK" ] || \
		error "symlink pathname ends with a slash"
	[ -n "$SYMLINK_TARGET" ] || error "original symlink target is missing"
	[ -n "$1" ] || error "maintainer script parameters are missing"

	debug "Executing $0 symlink_to_dir in $DPKG_MAINTSCRIPT_NAME" \
	      "of $DPKG_MAINTSCRIPT_PACKAGE"
	debug "SYMLINK=$SYMLINK -> $SYMLINK_TARGET PACKAGE=$PACKAGE" \
	      "LASTVERSION=$LASTVERSION ACTION=$1 PARAM=$2"

	case "$DPKG_MAINTSCRIPT_NAME" in
	preinst)
		if [ "$1" = "install" -o "$1" = "upgrade" ] &&
		   [ -n "$2" ] && [ -h "$SYMLINK" ] &&
		   symlink_match "$SYMLINK" "$SYMLINK_TARGET" &&
		   dpkg --compare-versions "$2" le-nl "$LASTVERSION"; then
			mv -f "$SYMLINK" "${SYMLINK}.dpkg-backup"
		fi
		;;
	postinst)
		# We cannot bail depending on the version, as here we only
		# know what was the last configured version, and we might
		# have been unpacked, then upgraded with an unpack and thus
		# never been configured before.
		if [ "$1" = "configure" ] && [ -h "${SYMLINK}.dpkg-backup" ] &&
		   symlink_match "${SYMLINK}.dpkg-backup" "$SYMLINK_TARGET"
		then
			rm -f "${SYMLINK}.dpkg-backup"
		fi
		;;
	postrm)
		if [ "$1" = "purge" ] && [ -h "${SYMLINK}.dpkg-backup" ]; then
		    rm -f "${SYMLINK}.dpkg-backup"
		fi
		if [ "$1" = "abort-install" -o "$1" = "abort-upgrade" ] &&
		   [ -n "$2" ] &&
		   [ ! -e "$SYMLINK" ] && [ -h "${SYMLINK}.dpkg-backup" ] &&
		   symlink_match "${SYMLINK}.dpkg-backup" "$SYMLINK_TARGET" &&
		   dpkg --compare-versions "$2" le-nl "$LASTVERSION"; then
			echo "Restoring backup of $SYMLINK ..."
			mv "${SYMLINK}.dpkg-backup" "$SYMLINK"
		fi
		;;
	*)
		debug "$0 symlink_to_dir not required in $DPKG_MAINTSCRIPT_NAME"
		;;
	esac
}

##
## Functions to replace a directory with a symlink
##
dir_to_symlink() {
	local PATHNAME="${1%/}"
	local SYMLINK_TARGET="$2"
	local LASTVERSION="$3"
	local PACKAGE="$4"

	if [ "$LASTVERSION" = "--" ]; then
		LASTVERSION=""
		PACKAGE="$DPKG_MAINTSCRIPT_PACKAGE${DPKG_MAINTSCRIPT_ARCH:+:$DPKG_MAINTSCRIPT_ARCH}"
	fi
	if [ "$PACKAGE" = "--" -o -z "$PACKAGE" ]; then
		PACKAGE="$DPKG_MAINTSCRIPT_PACKAGE${DPKG_MAINTSCRIPT_ARCH:+:$DPKG_MAINTSCRIPT_ARCH}"
	fi

	# Skip remaining parameters up to --
	while [ "$1" != "--" -a $# -gt 0 ]; do shift; done
	[ $# -gt 0 ] || badusage "missing arguments after --"
	shift

	[ -n "$DPKG_MAINTSCRIPT_NAME" ] || \
		error "environment variable DPKG_MAINTSCRIPT_NAME is required"
	[ -n "$PACKAGE" ] || error "cannot identify the package"
	[ -n "$PATHNAME" ] || error "directory parameter is missing"
	[ "${PATHNAME#/}" = "$PATHNAME" ] && \
		error "directory parameter is not an absolute path"
	[ -n "$SYMLINK_TARGET" ] || error "new symlink target is missing"
	[ -n "$1" ] || error "maintainer script parameters are missing"

	debug "Executing $0 dir_to_symlink in $DPKG_MAINTSCRIPT_NAME" \
	      "of $DPKG_MAINTSCRIPT_PACKAGE"
	debug "PATHNAME=$PATHNAME SYMLINK_TARGET=$SYMLINK_TARGET" \
	      "PACKAGE=$PACKAGE LASTVERSION=$LASTVERSION ACTION=$1 PARAM=$2"

	case "$DPKG_MAINTSCRIPT_NAME" in
	preinst)
		if [ "$1" = "install" -o "$1" = "upgrade" ] &&
		   [ -n "$2" ] &&
		   [ ! -h "$PATHNAME" ] && [ -d "$PATHNAME" ] &&
		   dpkg --compare-versions "$2" le-nl "$LASTVERSION"; then
			prepare_dir_to_symlink "$PACKAGE" "$PATHNAME"
		fi
		;;
	postinst)
		# We cannot bail depending on the version, as here we only
		# know what was the last configured version, and we might
		# have been unpacked, then upgraded with an unpack and thus
		# never been configured before.
		if [ "$1" = "configure" ] &&
		   [ -d "${PATHNAME}.dpkg-backup" ] &&
		   [ ! -h "$PATHNAME" ] && [ -d "$PATHNAME" ] &&
		   [ -f "$PATHNAME/.dpkg-staging-dir" ]; then
			finish_dir_to_symlink "$PATHNAME" "$SYMLINK_TARGET"
		fi
		;;
	postrm)
		if [ "$1" = "purge" ] && [ -d "${PATHNAME}.dpkg-backup" ]; then
		    rm -rf "${PATHNAME}.dpkg-backup"
		fi
		if [ "$1" = "abort-install" -o "$1" = "abort-upgrade" ] &&
		   [ -n "$2" ] &&
		   [ -d "${PATHNAME}.dpkg-backup" ] &&
		   [ \( ! -h "$PATHNAME" -a -d "$PATHNAME" -a \
		        -f "$PATHNAME/.dpkg-staging-dir" \) -o \
		     \( -h "$PATHNAME" -a \
		        \( "$(readlink $PATHNAME)" = "$SYMLINK_TARGET" -o \
		           "$(readlink -f $PATHNAME)" = "$SYMLINK_TARGET" \) \) ] &&
		   dpkg --compare-versions "$2" le-nl "$LASTVERSION"; then
			abort_dir_to_symlink "$PATHNAME"
		fi
		;;
	*)
		debug "$0 dir_to_symlink not required in $DPKG_MAINTSCRIPT_NAME"
		;;
	esac
}

prepare_dir_to_symlink()
{
	local PACKAGE="$1"
	local PATHNAME="$2"

	# If there are conffiles we should not perform the switch.
	if dpkg-query -W -f='${Conffiles}' "$PACKAGE" | \
	   grep -q "$PATHNAME/."; then
		error "directory '$PATHNAME' contains conffiles," \
		      "cannot switch to symlink"
	fi

	# If there are locally created files or files owned by another package
	# we should not perform the switch.
	find "$PATHNAME" -print0 | xargs -0 -n1 sh -c '
		package="$1"
		file="$2"
		if ! dpkg-query -L "$package" | grep -q -x "$file"; then
			exit 1
		fi
		exit 0
	' check-files-ownership "$PACKAGE" || \
		error "directory '$PATHNAME' contains files not owned by" \
		      "package $PACKAGE, cannot switch to symlink"

	# At this point, we know that the directory either contains no files,
	# or only non-conffiles owned by the package.
	#
	# To do the switch we cannot simply replace it with the final symlink
	# just yet, because dpkg needs to remove any file present in the old
	# package that have disappeared in the new one, and we do not want to
	# lose files resolving to the same pathname in the symlink target.
	#
	# We cannot replace the directory with a staging symlink either,
	# because dpkg will update symlinks to their new target.
	#
	# So we need to create a staging directory, to avoid removing files
	# from other packages, and to trap any new files in the directory
	# to move them to their correct place later on.
	mv -f "$PATHNAME" "${PATHNAME}.dpkg-backup"
	mkdir "$PATHNAME"

	# Mark it as a staging directory, so that we can track things.
	touch "$PATHNAME/.dpkg-staging-dir"
}

finish_dir_to_symlink()
{
	local PATHNAME="$1"
	local SYMLINK_TARGET="$2"

	# Move the contents of the staging directory to the symlink target,
	# as those are all new files installed between this package being
	# unpacked and configured.
	local ABS_SYMLINK_TARGET
	if [ "${SYMLINK_TARGET#/}" = "$SYMLINK_TARGET" ]; then
		ABS_SYMLINK_TARGET="$(dirname "$PATHNAME")/$SYMLINK_TARGET"
	else
		ABS_SYMLINK_TARGET="$SYMLINK_TARGET"
	fi
	rm "$PATHNAME/.dpkg-staging-dir"
	find "$PATHNAME" -mindepth 1 -print0 | \
		xargs -0 -i% mv -f "%" "$ABS_SYMLINK_TARGET/"

	# Remove the staging directory.
	rmdir "$PATHNAME"

	# Do the actual switch.
	ln -s "$SYMLINK_TARGET" "$PATHNAME"

	# We are left behind the old files owned by this package in the backup
	# directory, just remove it.
	rm -rf "${PATHNAME}.dpkg-backup"
}

abort_dir_to_symlink()
{
	local PATHNAME="$1"

	echo "Restoring backup of $PATHNAME ..."
	if [ -h "$PATHNAME" ]; then
		rm -f "$PATHNAME"
	else
		# The staging directory must be empty, as no other package
		# should have been unpacked inbetween.
		rm -f "$PATHNAME/.dpkg-staging-dir"
		rmdir "$PATHNAME"
	fi

	mv "${PATHNAME}.dpkg-backup" "$PATHNAME"
}

# Common functions
ensure_package_owns_file() {
	local PACKAGE="$1"
	local FILE="$2"

	if ! dpkg-query -L "$PACKAGE" | grep -q -x "$FILE"; then
		debug "File '$FILE' not owned by package " \
		      "'$PACKAGE', skipping $command"
		return 1
	fi
	return 0
}

symlink_match()
{
	local SYMLINK="$1"
	local SYMLINK_TARGET="$2"

	[ "$(readlink $SYMLINK)" = "$SYMLINK_TARGET" ] || \
	[ "$(readlink -f $SYMLINK)" = "$SYMLINK_TARGET" ]
}

debug() {
	if [ -n "$DPKG_DEBUG" ]; then
		echo "DEBUG: $PROGNAME: $*" >&2
	fi
}

error() {
	echo "$PROGNAME: error: $*" >&2
	exit 1
}

warning() {
	echo "$PROGNAME: warning: $*" >&2
}

usage() {
	cat <<END
Usage: $PROGNAME <command> <parameter>... -- <maintainer-script-parameter>...

Commands:
  supports <command>
	Returns 0 (success) if the given command is supported, 1 otherwise.
  rm_conffile <conffile> [<last-version> [<package>]]
	Remove obsolete conffile. Must be called in preinst, postinst and
	postrm.
  mv_conffile <old-conf> <new-conf> [<last-version> [<package>]]
	Rename a conffile. Must be called in preinst, postinst and postrm.
  symlink_to_dir <pathname> <old-symlink-target> [<last-version> [<package>]]
	Replace a symlink with a directory. Must be called in preinst,
	postinst and postrm.
  dir_to_symlink <pathname> <new-symlink-target> [<last-version> [<package>]]
	Replace a directory with a symlink. Must be called in preinst,
	postinst and postrm.
  help
	Display this usage information.
END
}

badusage() {
	echo "$PROGNAME: error: $1" >&2
	echo >&2
	echo "Use '$PROGNAME help' for program usage information." >&2
	exit 1
}

# Main code
set -e

PROGNAME=$(basename $0)
version="unknown"
command="$1"
[ $# -gt 0 ] || badusage "missing command"
shift

case "$command" in
supports)
	case "$1" in
	rm_conffile|mv_conffile|symlink_to_dir|dir_to_symlink)
		code=0
		;;
	*)
		code=1
		;;
	esac
	if [ -z "$DPKG_MAINTSCRIPT_NAME" ]; then
		warning "environment variable DPKG_MAINTSCRIPT_NAME missing"
		code=1
	fi
	if [ -z "$DPKG_MAINTSCRIPT_PACKAGE" ]; then
		warning "environment variable DPKG_MAINTSCRIPT_PACKAGE missing"
		code=1
	fi
	exit $code
	;;
rm_conffile)
	rm_conffile "$@"
	;;
mv_conffile)
	mv_conffile "$@"
	;;
symlink_to_dir)
	symlink_to_dir "$@"
	;;
dir_to_symlink)
	dir_to_symlink "$@"
	;;
--help|help|-?)
	usage
	;;
--version)
	cat <<-END
	Debian $PROGNAME version $version.

	This is free software; see the GNU General Public License version 2 or
	later for copying conditions. There is NO warranty.
	END
	;;
*)
	badusage "command $command is unknown
Hint: upgrading dpkg to a newer version might help."
esac

exit 0
